File: android/app/src/main/java/com/clock3/pet/service/PetService.kt
package com.clock3.pet.service

import android.content.Context
import android.os.Handler
import android.os.Looper
import com.clock3.pet.R
import com.clock3.pet.data.model.Pet
import com.clock3.pet.data.repository.Clock3Repository
import com.clock3.pet.utils.ExpCalculator
import com.clock3.pet.utils.AppLog
import kotlinx.coroutines.*
import kotlinx.coroutines.sync.Mutex
import kotlinx.coroutines.sync.withLock
import java.time.LocalDateTime
import java.util.concurrent.CopyOnWriteArrayList

class PetService private constructor(context: Context) {
    private val appContext = context.applicationContext
    private val repository = Clock3Repository(appContext)
    @Volatile
    private var pet: Pet? = null
    private val scope = CoroutineScope(Dispatchers.Default + SupervisorJob())
    private val mainHandler = Handler(Looper.getMainLooper())
    @Volatile
    private var isInitialized = false
    private val initMutex = Mutex()
    private var initDeferred = CompletableDeferred<Unit>()

    private val statusCallbacks = CopyOnWriteArrayList<(Pet) -> Unit>()

    init {
        scope.launch {
            try {
                pet = repository.getPetSync()
                isInitialized = true
                initDeferred.complete(Unit)
            } catch (e: Exception) {
                AppLog.e(TAG, "Failed to initialize PetService", e)
                isInitialized = true
                initDeferred.complete(Unit)
            }
        }
    }

    suspend fun loadPet(): Pet {
        if (!isInitialized) {
            initDeferred.await()
        }
        if (pet == null) {
            initMutex.withLock {
                if (pet == null) {
                    pet = repository.getPetSync()
                }
            }
        }
        if (pet == null) {
            val defaultPet = Pet()
            this.pet = defaultPet
            return defaultPet
        }
        return pet ?: Pet()
    }

    suspend fun savePet() {
        pet?.let {
            repository.savePet(it)
        }
    }

    suspend fun interact(): Map<String, Any> {
        val currentPet = pet ?: return mapOf("message" to appContext.getString(R.string.pet_not_loaded))
        val (updatedPet, levelUpMessages) = currentPet.interact()
        pet = updatedPet

        savePet()
        notifyStatusChange()

        return mapOf(
            "message" to updatedPet.getRandomMessage(),
            "mood" to updatedPet.mood.name,
            "emoji" to updatedPet.mood.emoji,
            "level" to updatedPet.level,
            "exp" to updatedPet.exp,
            "exp_needed" to updatedPet.getExpForNextLevel(),
            "level_up" to levelUpMessages
        )
    }

    suspend fun feedPet(): Map<String, Any> {
        val currentPet = pet ?: return mapOf("message" to appContext.getString(R.string.pet_not_loaded))
        val happyPet = currentPet.copy(mood = Pet.PetMood.HAPPY)
        val (updatedPet, levelUpMessages) = happyPet.addExp(ExpCalculator.calculateFeedExp())
        pet = updatedPet

        savePet()
        notifyStatusChange()

        return mapOf(
            "message" to appContext.getString(R.string.pet_feed_msg),
            "level_up" to levelUpMessages
        )
    }

    suspend fun playWithPet(): Map<String, Any> {
        val currentPet = pet ?: return mapOf("message" to appContext.getString(R.string.pet_not_loaded))
        val excitedPet = currentPet.copy(mood = Pet.PetMood.EXCITED)
        val (updatedPet, levelUpMessages) = excitedPet.addExp(ExpCalculator.calculatePlayExp())
        pet = updatedPet

        savePet()
        notifyStatusChange()

        return mapOf(
            "message" to appContext.getString(R.string.pet_play_msg),
            "level_up" to levelUpMessages
        )
    }

    suspend fun sleep(): Map<String, Any> {
        val currentPet = pet ?: return mapOf("message" to appContext.getString(R.string.pet_not_loaded))
        pet = currentPet.copy(mood = Pet.PetMood.SLEEPY)
        savePet()
        notifyStatusChange()

        return mapOf("message" to appContext.getString(R.string.pet_sleep_msg))
    }

    suspend fun addExp(amount: Int): Map<String, Any> {
        val currentPet = pet ?: return mapOf("message" to appContext.getString(R.string.pet_not_loaded))
        val safeAmount = amount.coerceAtLeast(0)
        val (updatedPet, levelUpMessages) = currentPet.addExp(safeAmount)
        val rows = repository.atomicUpdatePetExp(updatedPet.level, updatedPet.exp, currentPet.level, currentPet.exp)
        if (rows > 0) {
            pet = updatedPet
        } else {
            val freshPet = repository.getPetSync()
            val (retriedPet, retriedMessages) = freshPet.addExp(safeAmount)
            pet = retriedPet
            repository.savePet(retriedPet)
            return mapOf(
                "exp_added" to safeAmount,
                "level_up" to retriedMessages,
                "level" to retriedPet.level,
                "exp" to retriedPet.exp
            )
        }
        notifyStatusChange()

        return mapOf(
            "exp_added" to safeAmount,
            "level_up" to levelUpMessages,
            "level" to updatedPet.level,
            "exp" to updatedPet.exp
        )
    }

    fun isSleeping(): Boolean {
        return pet?.isSleeping() ?: false
    }

    fun getStatus(): Map<String, Any> {
        val currentPet = pet ?: return emptyMap()
        return mapOf(
            "name" to currentPet.name,
            "mood" to currentPet.mood.name,
            "emoji" to currentPet.mood.emoji,
            "level" to currentPet.level,
            "exp" to currentPet.exp,
            "exp_needed" to currentPet.getExpForNextLevel(),
            "is_sleeping" to currentPet.isSleeping(),
            "total_interactions" to currentPet.totalInteractions
        )
    }

    suspend fun updateName(name: String) {
        pet?.let {
            pet = it.copy(name = name)
            savePet()
            notifyStatusChange()
        }
    }

    fun onStatusChange(callback: (Pet) -> Unit) {
        statusCallbacks.add(callback)
    }

    fun removeStatusCallback(callback: (Pet) -> Unit) {
        statusCallbacks.remove(callback)
    }

    private fun notifyStatusChange() {
        pet?.let { currentPet ->
            mainHandler.post {
                statusCallbacks.forEach { callback ->
                    try {
                        callback(currentPet)
                    } catch (e: Exception) {
                        AppLog.e(TAG, "Status callback error", e)
                    }
                }
            }
        }
    }

    fun release() {
        scope.cancel()
        initDeferred.cancel()
        statusCallbacks.clear()
        pet = null
        isInitialized = false
    }

    fun reset() {
        scope.cancel()
        val newDeferred = CompletableDeferred<Unit>()
        initDeferred = newDeferred
        isInitialized = false
        pet = null
        val newScope = CoroutineScope(Dispatchers.Default + SupervisorJob())
        newScope.launch {
            try {
                pet = repository.getPetSync()
                isInitialized = true
                newDeferred.complete(Unit)
            } catch (e: Exception) {
                AppLog.e(TAG, "Failed to re-initialize PetService", e)
                isInitialized = true
                newDeferred.complete(Unit)
            }
        }
    }

    companion object {
        private const val TAG = "PetService"

        @Volatile
        private var INSTANCE: PetService? = null

        fun getInstance(context: Context): PetService {
            return INSTANCE ?: synchronized(this) {
                INSTANCE ?: PetService(context.applicationContext).also { INSTANCE = it }
            }
        }

        fun releaseInstance() {
            INSTANCE?.release()
            INSTANCE = null
        }
    }
}

File: app/services/pet_service.py
import random
from datetime import datetime
from typing import Callable, Optional
from app.database import db
from app.models.pet import Pet

class PetService:
    """宠物服务"""
    
    _instance = None
    
    def __new__(cls):
        if cls._instance is None:
            cls._instance = super().__new__(cls)
            cls._instance._initialized = False
        return cls._instance
    
    def __init__(self):
        if self._initialized:
            return
        
        self._initialized = True
        self.pet: Optional[Pet] = None
        self.callbacks = []
    
    def load_pet(self):
        """加载宠物数据"""
        self.pet = db.get_pet_data()
    
    def save_pet(self):
        """保存宠物数据"""
        if self.pet:
            db.save_pet_data(self.pet)
    
    def interact(self) -> dict:
        """与宠物交互"""
        if not self.pet:
            return {'message': '宠物未加载', 'level_up': []}
        
        level_up_messages = self.pet.interact()
        self.save_pet()
        
        return {
            'message': self.pet.get_random_message(),
            'mood': self.pet.mood,
            'emoji': self.pet.get_mood_emoji(),
            'level': self.pet.level,
            'exp': self.pet.exp,
            'exp_needed': self.pet.get_exp_for_next_level(),
            'level_up': level_up_messages
        }
    
    def feed_pet(self):
        """喂宠物"""
        if not self.pet:
            return {'message': '宠物未加载', 'level_up': []}
        
        self.pet.mood = 'happy'
        self.pet.last_interaction = datetime.now().isoformat()
        self.pet.total_interactions += 1
        exp_gain = self.pet.add_exp(10)
        self.save_pet()
        
        return {
            'message': '好吃! 谢谢喂我~',
            'mood': self.pet.mood,
            'emoji': self.pet.get_mood_emoji(),
            'level': self.pet.level,
            'exp': self.pet.exp,
            'exp_needed': self.pet.get_exp_for_next_level(),
            'level_up': exp_gain
        }
    
    def play_with_pet(self):
        """和宠物玩"""
        if not self.pet:
            return {'message': '宠物未加载', 'level_up': []}
        
        self.pet.mood = 'excited'
        self.pet.last_interaction = datetime.now().isoformat()
        self.pet.total_interactions += 1
        exp_gain = self.pet.add_exp(15)
        self.save_pet()
        
        return {
            'message': '太好玩了! 再来再来!',
            'mood': self.pet.mood,
            'emoji': self.pet.get_mood_emoji(),
            'level': self.pet.level,
            'exp': self.pet.exp,
            'exp_needed': self.pet.get_exp_for_next_level(),
            'level_up': exp_gain
        }
    
    def pet_sleep(self):
        """宠物睡觉"""
        if not self.pet:
            return {'message': '宠物未加载'}
        
        self.pet.mood = 'sleepy'
        self.save_pet()
        
        return {
            'message': '困了... zzZ...',
            'mood': self.pet.mood,
            'emoji': self.pet.get_mood_emoji(),
            'level': self.pet.level,
            'exp': self.pet.exp,
            'exp_needed': self.pet.get_exp_for_next_level()
        }
    
    def is_sleeping(self) -> bool:
        """检查宠物是否在睡眠"""
        if self.pet:
            return self.pet.is_sleeping()
        return False
    
    def get_status(self) -> dict:
        """获取宠物状态"""
        if not self.pet:
            return {}
        
        return {
            'name': self.pet.name,
            'mood': self.pet.mood,
            'emoji': self.pet.get_mood_emoji(),
            'level': self.pet.level,
            'exp': self.pet.exp,
            'exp_needed': self.pet.get_exp_for_next_level(),
            'is_sleeping': self.pet.is_sleeping(),
            'total_interactions': self.pet.total_interactions
        }
    
    def update_name(self, name: str):
        """更新宠物名字"""
        if self.pet:
            self.pet.name = name
            self.save_pet()
    
    def on_status_change(self, callback: Callable):
        """注册状态变化回调"""
        self.callbacks.append(callback)
    
    def notify_status_change(self):
        """通知状态变化"""
        status = self.get_status()
        for callback in self.callbacks:
            try:
                callback(status)
            except Exception as e:
                print(f"宠物状态回调错误: {e}")

pet_service = PetService()
